### PR TITLE
Fixes evil-mode dragging and adds evil-mode recipes

### PR DESCRIPTION
**Pull request by gabrielsimoes**
This uses @luxbock fix from https://github.com/rejeep/drag-stuff.el/pull/10 and adds recipes for testing evil-mode.

### CODE REVIEW

**Comment by gabrielsimoes** [comment]
Well, all checks were working on my machine (emacs 24.5.1), I don't think it's evil-mode's fault.

**Comment by gabrielsimoes** [comment]
Fixed, checks are working now

**Comment by gabrielsimoes** [comment]
Any update? Please don't let this project die.

File: features/conflicting-modes.feature [hunk 1]
```diff
@@ -96,3 +96,51 @@ Feature: Drag Stuff
     And I turn on drag-stuff
     And I drag line "1" down
     Then longlines-mode should not be active
+
+  Scenario Outline: Evil region left/right
+	Given I insert:
```
Comment (rejeep) [inline features/conflicting-modes.feature]: Invalid indentation

File: drag-stuff.el [hunk 2]
```diff
@@ -183,9 +187,12 @@
...
 (defun drag-stuff-lines-vertically (fn)
   "Yields variables used to drag lines vertically."
-  (let* ((deactivate-mark nil)
+  (let* ((evilp (drag-stuff--evil-p))
+         (vis-type (when evilp (evil-visual-type)))
+         (deactivate-mark nil)
          (mark-line (line-number-at-pos (mark)))
-         (point-line (line-number-at-pos (point)))
+         (point-line (line-number-at-pos
+                      (if evilp (evil-visual-goto-end) (point))))
```
Comment (rejeep) [inline drag-stuff.el]: Can you please explain this change as there is a side effect in evil, but not in default.

**Comment by rejeep** [comment]
@gabrielsimoes Really sorry about this. I don't know how, but I didn't notice this pull request when it was first sent.

I made a few comments in the code. Can you please address them and when done squash your commits into a single one.

**Comment by gabrielsimoes** [comment]
No problem, sorry if it seemed rude or anything. I'll try to address your comment as soon as I can.

File: drag-stuff.el (hunk 2 above)
Comment (gabrielsimoes) [inline drag-stuff.el]: I still have to read the code, actually I just got the changes from @luxbock pull request and added test recipes. However, I noticed evil-mode also selects the lower line each time you drag up or down. So it ends up draggin more than it should every time. This change fixes it, but makes it unable to drag down to the last line of the file. I will try to provide a better solution.

File: drag-stuff.el [hunk 3]
```diff
@@ -167,7 +171,7 @@
 
 (defun drag-stuff-lines-up (arg)
   "Move all lines in the selected region ARG lines up."
-  (if (> (line-number-at-pos (min (point) (mark))) (abs arg))
+  (if (> (line-number-at-pos (region-beginning)) (abs arg))
```
Comment (gabrielsimoes) [inline drag-stuff.el]: I thought this was a better solution, if you prefer the other way, I may roll back this change.

File: drag-stuff.el [hunk 4]
```diff
@@ -175,22 +179,34 @@
 
 (defun drag-stuff-lines-down (arg)
   "Move all lines in the selected region ARG lines up."
-  (if (<= (+ (line-number-at-pos (max (point) (mark))) arg) (count-lines (point-min) (point-max)))
+  (let ((lastline (if (and (drag-stuff--evil-p)
```
Comment (gabrielsimoes) [inline drag-stuff.el]: Evil visual selection mode has a strange behaviour: If you press <End> it actually doesn't move to the last char of the line, but the first char of the next line. So, this checks if this is happening, and goes back one char before testing.

**Comment by gabrielsimoes** [comment]
I thought I had fixed it, but some new bugs appeared.

**Comment by gabrielsimoes** [comment]
Alright, I wrote a better solution and squashed the commits. You may want to change something.

File: drag-stuff.el [hunk 5]
```diff
@@ -167,30 +171,44 @@
...
+  (if (> (line-number-at-pos (region-beginning)) (abs arg))
       (drag-stuff-lines-vertically
        (lambda (beg end)
          (drag-stuff-drag-region-up beg end arg)))
     (message "Can not move lines further up")))
 
 (defun drag-stuff-lines-down (arg)
   "Move all lines in the selected region ARG lines up."
-  (if (<= (+ (line-number-at-pos (max (point) (mark))) arg) (count-lines (point-min) (point-max)))
+  (let ((lastline (if (drag-stuff--evil-p)
```
Comment (rejeep) [inline drag-stuff.el]: The name `lastline` is a bit misleading as it's not a line. It's the point at the end of the region. So let's call it something like `region-end` or similar.

File: drag-stuff.el [hunk 6]
```diff
@@ -167,30 +171,44 @@
...
+         (point-line (line-number-at-pos (if (and evilp (> (point) (mark)))
+                                           (save-mark-and-excursion (evil-visual-goto-end))
+                                           (point))))
+         (mark-col (save-mark-and-excursion (if (and evilp (< (point) (mark)))
+                                              (evil-visual-goto-end)
+                                              (exchange-point-and-mark))
+                                            (current-column)))
+         (point-col (if (and evilp (> (point) (mark)))
+                      (save-mark-and-excursion (evil-visual-goto-end) (current-column))
+                      (current-column)))
```
Comment (rejeep) [inline drag-stuff.el]: I think all these should be in private (functions starting with `drag-stuff--`) functions. For example:

* `drag-stuff--mark-line`
* `drag-stuff--point-line`
* `drag-stuff--mark-col`
* `drag-stuff--point-col`

File: drag-stuff.el [hunk 7]
```diff
@@ -225,12 +247,17 @@
 (defun drag-stuff-whole-lines-region ()
   "Return the positions of the region with whole lines included."
   (let (beg end)
-    (if (> (point) (mark))
-        (exchange-point-and-mark))
-    (setq beg (line-beginning-position))
-    (if mark-active
-        (exchange-point-and-mark))
-    (setq end (line-end-position))
+    (if evilp
```
Comment (rejeep) [inline drag-stuff.el]: Use `cond`

File: drag-stuff.el [hunk 8]
```diff
@@ -225,12 +247,17 @@
...
+      (progn (setq beg (save-mark-and-excursion (goto-char (region-beginning)) (line-beginning-position)))
+             (setq end (save-mark-and-excursion (evil-visual-goto-end) (line-end-position))))
```
Comment (rejeep) [inline drag-stuff.el]: As we want both beg and end, I think it would make sense to use `evil-visual-range` here instead.

Comment (gabrielsimoes) [inline drag-stuff.el]: Didn't know about it, but actually `evil-visual-range` is affected by evil's behaviour on selecting one more char than it should when selecting the end of a line, but `evil-visual-goto-end` isn't.

**Comment by gabrielsimoes** [comment]
Great points, I also aligned let* declarations at `drag-stuff-lines-vertically`.

File: drag-stuff.el [hunk 9]
```diff
@@ -167,30 +195,35 @@
...
+  (let* ((evilp      (drag-stuff--evil-p))
+         (mark-line  (drag-stuff--mark-line))
+         (point-line (drag-stuff--point-line))
+         (mark-col   (drag-stuff--mark-col))
+         (point-col  (drag-stuff--point-col))
+         (bounds     (drag-stuff-whole-lines-region))
+         (beg        (car bounds))
+         (end        (car (cdr bounds)))
+         (deactivate-mark nil))
+
```
Comment (rejeep) [inline drag-stuff.el]: Please don't add whitespace like this. As you can see, it's impossible to see what changed in this diff.

File: drag-stuff.el [hunk 10]
```diff
@@ -89,13 +89,41 @@
...
+  (and (bound-and-true-p evil-mode) (evil-visual-state-p)))
+
 (defun drag-stuff--kbd (key)
   "Key binding helper."
   (let ((mod (if (listp drag-stuff-modifier)
                  drag-stuff-modifier
                (list drag-stuff-modifier))))
     (vector (append mod (list key)))))
 
+(defun drag-stuff--mark-line ()
```
Comment (rejeep) [inline drag-stuff.el]: I know it was my idea, but can you change the name of the functions to:

* `drag-stuff--line-at-mark`
* `drag-stuff--line-at-point`
* `drag-stuff--col-at-mark`
* `drag-stuff--col-at-point`

I didn't realise att first that the name `drag-stuff--mark-line` sound like it marks a line. :smile:

**Comment by gabrielsimoes** [comment]
Ok, I renamed the functions, and rewrote them with better evil's variables for `point` and `mark`. With this rewrite, and also an addition at the end of `drag-stuff-lines-vertically`, the cursor location after dragging lines with evil's line selection is fixed (just a minor bug). I also removed the spaces. I won't ammend the commits until merging this.

**Comment by gabrielsimoes** [comment]
I plan on adding tests for the cursor position later, but I think it is better to keep it for another pull request.

**Comment by rejeep** [comment]
> I plan on adding tests for the cursor position later, but I think it is better to keep it for another pull request.

Then the feature should be in a separate PR as well. So either add the test in this PR or fix that bug (with test) in another PR.

**Comment by gabrielsimoes** [comment]
Sure, I prefer adding the tests here. But I will only add tests for evil mode's behaviour, and keep cursor tests for emacs in general out of this PR.

**Comment by gabrielsimoes** [comment]
I think I got a little confused on my last two comments here. :sweat_smile: Anyway, I wrote tests for all the things evil-mode might behave differently and are fixed in this PR, and placed everything at `evil-mode.feature`.

**Comment by rejeep** [comment]
I will try out this branch locally for a little while and if everything works, I'll merge it.

**Comment by rejeep** [comment]
Merged, thanks for your work (and patience) on this! 👍

**Comment by gabrielsimoes** [comment]
Great! You're welcome

**Comment by gabrielsimoes** [comment]
This closes #10 and #5

**Pull request merged**